core/src/main/java/org/apache/iceberg/RemoveSnapshots.java

```diff
@@ -190,6 +282,10 @@ public void commit() {
         .onlyRetryOn(CommitFailedException.class)
         .run(item -> {
           TableMetadata updated = internalApply();
+          if (cleanExpiredFiles && (updated.refs().size() > 1 || multipleLineages(updated))) {
```

Comment:
The check is done after the computation of snapshots to remove but before the commit. I think this has to be done, since we need the updated commit graph after what the expiration would look like if we were to commit it to determine if there will be any detached lineages. If the check is done before the computation the procedure would miss some cases.